---
title: "Testing how much deterioration is needed to not find the closest copy-number profile"
author:
  - Barbara Hernando, bhernando@cnio.es
  - Blas Chaves, bchavesu@ext.cnio.es
  - Geoff Macintyre, gmacintyre@cnio.es
date: "`r format(Sys.time(), '%d %B, %Y')`"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{AssessingTool}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r setup, include = FALSE}
knitr::opts_chunk$set(collapse = TRUE,comment = "#>")
knitr::opts_chunk$set(echo = TRUE)
knitr::opts_chunk$set(message = FALSE)
knitr::opts_chunk$set(warning = FALSE)

## PACKAGES NEEDED
library(dplyr)
library(CNpare)
library(ggplot2)
library(ggthemes)

## DATA DIRECTORY
path_to_data<-paste0(dirname(rstudioapi::getSourceEditorContext()$path),"/data/")
```

### Testing the level of artificial perturbations needed to fail

Sequencing methods can introduce artefacts that deteriorate the copy number profile of a sample. This Rmarkdown is used to check how much deterioration is needed to not find the matched cell line by the **CNpare Tool**. For this purpose, we randomly removed bins of the copy number profiles derived from CCLE and tested the ability of CNpare to identify the matched cell line profiled as part of the GDSC project.

#### Step 1: set up to run the quality tool

##### [Input data]{.ul}

We first get the **RData** files in the `data/` folder:

-   `cells_segcn.RData` -- Absolute copy-number profiles of cancer cell lines generated with *ASCAT*. The segment tables should have the following column headers: "chromosome", "start", "end", "segVal", "sample".
-   `cells_mapping.RData` -- Mapping information of cancer cell lines

```{r input, include=FALSE}
cells_segcn<-cells_segcn
cells_mapping<-cells_mapping
```

##### [Get the genomic positions of bins]{.ul}

Segmented copy-number profiles of cell lines may be fitted to the genomic positions of bins. Therefore, we split genome in evenly sized bins (default 500 kb) and get their positions.

```{r pos_bins}
allchr=c(1:22) #Add 23 if you want to include chrX
lengthChr=lengthChr
posBins <- lapply(allchr,function(chr) 
    getBinsStartsEnds(window=500000, chr, lengthChr[chr]))
```

##### [Selection of cell lines with CIN]{.ul}

We selected cell lines that have at least 20 copy number aberrations since this tool has been developed to work with samples with a minimal level of chromosomal instability.

```{r select_samples, echo=FALSE}
cells_segcn<-getCINProfiles(segcn=cells_segcn, samples=unique(cells_segcn$sample))
print(paste0("Number of cell lines with detectable CIN: ",length(unique(cells_segcn$sample))))
```

##### [Cell lines data preprocessing]{.ul}

We selected cell lines included in both databases (CCLE and GDSC), and whose copy-number profile has been generated using data from both CCLE and GDSC. Then, *segment tables* are converted to *bin tables*. A matrix with bin-level copy numbers is generated for both CCLE and GDSC.

```{r cnbins}
#cell lines with CN profile generated with CCLE and GDSC data
ccle <- cells_mapping[cells_mapping$study=="CCLE",c(1,4)]
ccle <- ccle[ccle$fileid %in% cells_segcn$sample,] #with ASCAT CN profile
gdsc <- cells_mapping[cells_mapping$study=="GDSC",c(1,4)] #in GDSC cellid and fileid is the same
gdsc <- gdsc[gdsc$fileid %in% cells_segcn$sample,] #with ASCAT CN profile
both <- ccle[ccle$cellid %in% gdsc$fileid,] 

# Get copy-number data per bin
ccle_cn <- getCNbins(posBins=posBins, data=cells_segcn, samples=both$fileid)
gdsc_cn <- getCNbins(posBins=posBins, data=cells_segcn, samples=both$cellid)
```

#### Step 2: Testing the level of artificial perturbations needed to fail

We randomly introduce noise in a % of the segments in the copy-number profiles generated by ASCAT using CCLE data. Then, we compared get bin-level copy number profiles and compare them with the bin-level copy-number profiles of the same cell lines generated from the GDSC project. 

We tested the capacity of CNpare to identify the correct cell line after introducing noise. Manhattan distance was computed for profile comparisons since it's a measure affected by ploidy shifts.

##### 2.1. Perturbation of segment tables
```{r}
set.seed(123)

# Get segment tables of CCLE
ccle_segcn <- cells_segcn[cells_segcn$sample %in% both$fileid,]

# Add perturbations to segment tables
levels=c(10,20,30,40,50,60,70,80,90)

allPerturbations<-list()

allPerturbations<-lapply(1:10,function(x){
  perturbations<-list()
  
  perturbations<-lapply(1:length(levels), function(l){
    level.perturbation<-c()
    for (s in 1:length(both$fileid)){
      ccle_cn_perturbed=ccle_segcn[ccle_segcn$sample==both$fileid[s],]
      nsegs=nrow(ccle_cn_perturbed)
      
      #set of randomly selected bins to mask
      segs=sample(1:nsegs, round(levels[l]*nsegs/100), replace=FALSE)
      ccle_cn_perturbed[segs,"segVal"]=ccle_cn_perturbed[segs,"segVal"] + 
        sample(c(-1,1), length(segs), replace=TRUE)
    
      level.perturbation<-rbind(level.perturbation,ccle_cn_perturbed)
    }
    perturbations[[l]]<-level.perturbation
    
  })
  allPerturbations[[x]]<-perturbations
})

# Get bin tables
allPerturbed.bins <- list()

allPerturbed.bins<-lapply(1:10, function(x){
  perturbations<-allPerturbations[[x]]
  perturbed.bincn <- list()
  
  perturbed.bincn<-lapply(1:length(levels), function(l){
    ccle_bincn<-c()
    dat<-as.data.frame(perturbations[[l]])
    ccle_bincn<-getCNbins(posBins=posBins, data=dat, samples=both$fileid)
    perturbed.bincn[[l]]<-ccle_bincn
  })
  allPerturbed.bins[[x]]<-perturbed.bincn
})
#saveRDS(allPerturbed.bins, paste0(path_to_data, "/ccle_bins500kb_perturbated_allruns.rds"))
```

##### 2.2. Compute similarities between CCLE perturbed and GDSC profiles
```{r}
# Test perturbations
testPerturbations <- list()

testPerturbations <- lapply(1:10, function(x){
  perturbed.bincn<-allPerturbed.bins[[x]]
  perturbed.bincn[[10]]<-as.matrix(ccle_cn)
  names(perturbed.bincn)<-c(levels,0)
  
  tests <- list()
  mat.perturbations <- c()
  matches.m <- c()
  perturbations <- c()
  for(l in names(perturbed.bincn)){
    ccle_cn_perturbed <- perturbed.bincn[[l]]
    
    ##Manhattan similarities
    measures<-getSimilarities(dat1=ccle_cn_perturbed, dat2=gdsc_cn, method="manhattan")
    colnames(measures)<-c("fileid", "id","manhattan")
    measures<-left_join(measures,both,by="fileid")
    list.tophits<-getTopHit(samples=both$cellid, measure=measures[,c(4,2:3)], method="manhattan")
  
    ##Performance of matches
    list.tophits$equal<-list.tophits[,1]==list.tophits[,2]
    #empirical pvalue
    list.tophits<-left_join(list.tophits,measures,by=c("cellid","id"))
    mes <- measures[measures$id!=measures$cellid,]
    for(s in both$cellid){
      dist <- list.tophits$manhattan[list.tophits$cellid==s]
      n <- Position(function(x) x > dist, CNpare:::all_manhattan)-1
      p <- n/length(CNpare:::all_manhattan)
      list.tophits$pvalue[list.tophits$cellid==s]<-p
    }
    #significance of matches
    list.tophits$sign<-list.tophits$equal==TRUE & list.tophits$pvalue<0.05
    top1.sign<-nrow(list.tophits[list.tophits$equal & list.tophits$sign,])
    top1.nonsign<-nrow(list.tophits[list.tophits$equal & !list.tophits$sign,])
    nontop.sign<-nrow(list.tophits[!list.tophits$equal & list.tophits$sign,])
    nontop.nonsign<-nrow(list.tophits[!list.tophits$equal & !list.tophits$sign,])
    res<-c(top1.sign,top1.nonsign,nontop.sign,nontop.nonsign)
    
    #output
    matches.m<-rbind(matches.m,res)
  }
  mat.perturbations <- as.data.frame(matches.m)
  colnames(mat.perturbations) <- c("sign.match","nonsig.match","sign.nonmatch","nonsign.nonmatch")
  mat.perturbations$levels <- names(perturbed.bincn)

  testPerturbations[[x]]<-mat.perturbations
})
#saveRDS(testPerturbations, paste0(path_to_data, "/ccle_bins500kb_perturbated_allruns_manhattan.rds"))
```

##### 2.3. Plotting results
```{r}
#Prepare data
testPerturbations <- data.table::rbindlist(testPerturbations)
dt <- reshape2::melt(testPerturbations)
dt <- dt %>% group_by(variable,levels) %>% summarise(median=median(value),min=min(value),max=max(value))
```

```{r}
##Plotting
#remove significant non-matched by CNpare since it's 0 in all perturbation levels
dt <- dt[dt$variable!="sign.nonmatch",]
#prepare variables to plot
dt$variable <- factor(dt$variable, levels =c("nonsign.nonmatch","sign.match","nonsig.match"),
                      labels = c("Non-significant incorrect match",
                                 "Significant correct match", "Non-significant correct match"))
dt$levels <- factor(dt$levels, levels = c("0","10","20","30","40","50","60","70","80","90"))

#prepare bars
dt$bar[dt$variable=="Significant correct match"] <- round(dt$median[dt$variable=="Significant correct match"],0)
dt$bar[dt$variable=="Non-significant incorrect match"] <- round(dt$median[dt$variable=="Non-significant incorrect match"],0)
dt$bar[dt$variable=="Non-significant correct match"] <- 304-(dt$bar[dt$variable=="Non-significant incorrect match"]+dt$bar[dt$variable=="Significant correct match"])

#prepare errors for stacked bar plot
dt$ci.min[dt$variable=="Non-significant correct match"] <- dt$min[dt$variable=="Non-significant correct match"]
dt$ci.max[dt$variable=="Non-significant correct match"] <- dt$max[dt$variable=="Non-significant correct match"]
dt$ci.min[dt$variable=="Significant correct match"] <- dt$min[dt$variable=="Significant correct match"] + dt$min[dt$variable=="Non-significant correct match"]
dt$ci.max[dt$variable=="Significant correct match"] <- dt$max[dt$variable=="Significant correct match"] + dt$max[dt$variable=="Non-significant correct match"]

#plot
p <- ggplot(dt, aes(x=levels, y=bar, fill=variable)) + 
  geom_bar(stat="identity") +
  geom_errorbar(aes(ymin=ci.min,ymax=ci.max), width=0.2, position="identity") +
  ylim(0,310)+
  xlab("% of noise") +
  ylab("number of cell lines") +
  scale_fill_manual(values=c("#f3e967","#4cbb9d","#e18e4c")) + #yellow #green #orange  
  theme_classic() +
  theme(text = element_text(size = 10),
        axis.text = element_text(size = 8),
        axis.line = element_line(size = 0.6), 
        axis.ticks = element_line(size = 0.6),
        axis.ticks.length = unit(.1, "cm"),
        legend.position = "none")

ggsave(paste0(path_to_data,"/SuppFig_4_Perturbations.png"), width = 120, height = 80, units="mm")
print(p)
```
