---
title: "Demonstrating why is useful to compare at genome-wide level"
author:
  - Barbara Hernando, bhernando@cnio.es
  - Blas Chaves, bchavesu@ext.cnio.es
  - Geoff Macintyre, gmacintyre@cnio.es
date: "`r format(Sys.time(), '%d %B, %Y')`"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{DemonstratingUtility}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(collapse = TRUE,comment = "#>")
knitr::opts_chunk$set(echo = TRUE)
knitr::opts_chunk$set(message = FALSE)
knitr::opts_chunk$set(warning = FALSE)

## PACKAGES
library(dplyr)
library(CNpare)

# DATA DIRECTORY
path_to_data<-"C:/Users/bhernando/Desktop/CNIO/Projects/BlasProject/quality_tool/data/"
```

### Finding the best model of chromosomal instability

This **CNpare Tool** aims at finding the best cell line model for studying a type of chromosomal instability (CIN) in vitro. 
This tool perform comparisons based on genome-wide copy number profiles. However, the majority of previous approaches perform DNA copy number based comparison at gene-level resolution.

This Rmarkdown aims at demonstrating the utility to use genome-wide copy number comparisons.

Here, we compared the next most representative cell line model for all cancer cell lines from the Cancer Cell Line Encyclopedia (CCLE) project using genome-wide and gene-level resolutions.

#### 1) Input data

We first get the **RData** files in the `data/` folder:

-   `cells_segcn.RData` -- Absolute copy-number profiles of cancer cell lines generated with *ASCAT*. The segment tables should have the following column headers: "chromosome", "start", "end", "segVal", "sample".
-   `cells_mapping.RData` Mapping information of cancer cell lines
-   `CCLE_metadata.RData` Metadata of CCLE database for selecting cell lines by tissue origin

```{r input, include=FALSE}
cells_segcn<-cells_segcn
cells_mapping<-cells_mapping
CCLE_metadata<-CCLE_metadata
CCLE_metadata$Cell.line.primary.name<-toupper(CCLE_metadata$Cell.line.primary.name)
```

#### 2) Comparison of copy number profiles

##### 2.1. [Genome-wide resolution]{.ul}

Segmented copy-number profiles of cell lines may be fitted to the genomic positions of bins. Therefore, we split genome in evenly sized bins (default 500 kb) and get their positions. 
```{r pos_bins}
bin.size=500 #in Kb
allchr=c(1:22) #Add 23 if want to include chrX
lengthChr<-lengthChr
posBins <- lapply(allchr,function(chr) 
    getBinsStartsEnds(window=bin.size*1000, chr, lengthChr[chr]))
```

We selected the CCLE models with available copy-number profile. We also selected profiles with detectable CIN (those with at least 20 CNAs in a profile). Therefore, we discarded profiles with non detectable chromosomal instability. Finally, a matrix with copy-numbers per bin is then generated.

```{r cnbins}
ccle<-as.data.frame(cbind(cellid=cells_mapping$cellid[cells_mapping$study=="CCLE"],sample=cells_mapping$fileid[cells_mapping$study=="CCLE"]))
cells_segcn <- cells_segcn[cells_segcn$sample%in%ccle$sample,]
cells_segcn <- getCINProfiles(segcn=cells_segcn, samples=unique(cells_segcn$sample))
cells_segcn <- left_join(cells_segcn,ccle, by="sample")[,c(1:4,6)]
colnames(cells_segcn)[5]<-"sample"

#get copy-number data per bin
ccle_cn <- getCNbins(posBins=posBins, data=cells_segcn, samples=unique(cells_segcn$sample))
```

We compare all CCLE profiles using four similarity metrics
```{r comparisons, warning=FALSE}
genome.measures<-getSimilarities(dat1=ccle_cn, dat2=ccle_cn, method="manhattan")
genome.measures<-genome.measures[!genome.measures[,1]==genome.measures[,2],]
```

##### 2.2. [Gene-level resolution]{.ul}

Gene positions were downloaded from UCSC Table Browser. We extracted the copy numbers in the transcription start of 18,944 protein-coding genes, and then compared gene-level copy numbers across cell lines using all similarity metrics. We also performed the analysis using only the list of cancer genes included in COSMIC (521 genes).
```{r}
## All UCSC genes
#gene positions from UCSC Table Browser
genes = read.table(paste0(path_to_data,"knownGenes_positions.txt"),sep="\t")[,c(1:4,7)]
genes = genes[!duplicated(genes$V3),] #no duplicates in start position
genes = genes[!duplicated(genes$V4),] #only one per protein name
genes = genes[!duplicated(genes$V7),] #only one per gene name --> randomly because positions are similar & we are using segmented tables to get the CN values

pos.genes<-genes[,c(5,2,3,3)] #column 3 txStart
colnames(pos.genes)<- c("name", "chromosome", "start", "end")
pos.genes$chromosome = substr(pos.genes$chromosome, 4, stop = 5)
chr<-c(1:22, "X")
pos.genes<-data.table::as.data.table(pos.genes[pos.genes$chromosome %in% chr,]) #remove genes in Y

#ccle
samples=unique(cells_segcn$sample)
names=pos.genes$name
ccle <- c()
for (s in 1:length(samples)){
    dat<-data.table::as.data.table(cells_segcn[cells_segcn$sample==samples[s],])
    dat<-dat[,c(5,4,1:3)]
    data.table::setkey(pos.genes, chromosome, start, end)
    overlap<-data.table::foverlaps(dat,pos.genes)
    overlap<-overlap[!is.na(overlap$name),]
    
    ccle<-rbind(ccle,overlap[,c(2,5,6)])
}

mccle<-reshape2::acast(ccle, name~sample, value.var="segVal")

#Comparisons at gene-level
gene.measures<-getSimilarities(dat1=mccle, dat2=mccle, method="manhattan")
gene.measures<-gene.measures[!gene.measures[,1]==gene.measures[,2],]

## COSMIC genes
#filter out non-cosmic genes
cosmic_genes = read.table(paste0(path_to_data,"COSMICgenes.txt"), fill = TRUE, sep="\t")[,1]
cgenes = genes[genes$V7 %in% cosmic_genes,] #only oncogenic genes

pos.genes<-cgenes[,c(5,2,3,3)] #column 3 txStart
colnames(pos.genes)<- c("name", "chromosome", "start", "end")
pos.genes$chromosome = substr(pos.genes$chromosome, 4, stop = 5)
chr<-c(1:22, "X")
pos.genes<-data.table::as.data.table(pos.genes[pos.genes$chromosome %in% chr,]) #remove genes in Y

#ccle
samples=unique(cells_segcn$sample)
names=pos.genes$name
ccle <- c()
for (s in 1:length(samples)){
    dat<-data.table::as.data.table(cells_segcn[cells_segcn$sample==samples[s],])
    dat<-dat[,c(5,4,1:3)]
    data.table::setkey(pos.genes, chromosome, start, end)
    overlap<-data.table::foverlaps(dat,pos.genes)
    overlap<-overlap[!is.na(overlap$name),]
    
    ccle<-rbind(ccle,overlap[,c(2,5,6)])
}

mccle<-reshape2::acast(ccle, name~sample, value.var="segVal")


#Comparisons at gene-level
cgene.measures<-getSimilarities(dat1=mccle, dat2=mccle, method="manhattan")
cgene.measures<-cgene.measures[!cgene.measures[,1]==cgene.measures[,2],]
```


#### 3) Comparison of resolution levels

Using manhattan distance
```{r}
genome.tophits<-getTopHit(samples=unique(cells_segcn$sample), measure=genome.measures, method="manhattan")
cgene.tophits<-getTopHit(samples=unique(cells_segcn$sample), measure=cgene.measures, method="manhattan")
gene.tophits<-getTopHit(samples=unique(cells_segcn$sample), measure=gene.measures, method="manhattan")
manhattan<-left_join(genome.tophits,cgene.tophits,by="cellid")
manhattan<-left_join(manhattan,gene.tophits,by="cellid")
colnames(manhattan)<-c("cellid","genome.match","cosmic.match","gene.match")
manhattan$gene.equal<-manhattan$genome.match==manhattan$gene.match
manhattan$cosmic.equal<-manhattan$genome.match==manhattan$cosmic.match

print("Number of discordant matches between genome-wide and 18k-gene-level resolutions") 
print(paste0(length(manhattan$gene.equal[!manhattan$gene.equal==TRUE]),"/604 samples"))

print("Number of discordant matches between genome-wide and cosmic-gene-level resolutions") 
print(paste0(length(manhattan$cosmic.equal[!manhattan$cosmic.equal==TRUE]),"/604 samples"))
```

Save results:
```{r}
write.table(manhattan, file=paste0(path_to_data,"ComparingToGeneLevel_manhattan_",bin.size,"kb.txt"), sep = "\t", quote = FALSE, col.names = TRUE, row.names = FALSE)
```


#### 4) Testing which resolution finds the best next representative model

Cosine similarity of signature activities between matched cell models at each resolution level (genome-wide, 18k genes and cosmic genes)
```{r}
# manhattan <- read.table(file=paste0(path_to_data,"ComparingToGeneLevel_manhattan_",bin.size,"kb.txt"), 
#                         sep = "\t", header = TRUE)

#signature quantification
component_parameters<-readRDS(paste0(path_to_data,"component_parameters.rds"))
cell.profiles <- getProfiles(segcn=cells_segcn, samples=unique(cells_segcn$sample))
CN_features <- extractCopynumberFeatures(cell.profiles)
sample_by_component <- generateSampleByComponentMatrix(CN_features, all_components=component_parameters)
signature_quantification <- quantifySignatures(sample_by_component)

#cosine 
samples=unique(cells_segcn$sample)
signatures.cosine <- list()
for (s in samples){
  matches=unique(as.character(manhattan[manhattan$cellid==s,c(2:4)]))
  
  cos <- lapply(matches, function(thisMatch) pair_cosine(cell=signature_quantification[,s],
                                                         ccle=signature_quantification[,thisMatch],
                                                         ccle.name=thisMatch,
                                                         pvalue=FALSE))
  cos <- do.call(rbind,cos)
  signatures.cosine[[s]] <- as.data.frame(cos)
}

signatures.cosine <- data.table::rbindlist(signatures.cosine, idcol=TRUE)
colnames(signatures.cosine) <- c("cellid", "match", "cos.sim")
signatures.cosine$cos.sim <- as.numeric(signatures.cosine$cos.sim)

#merge genome-wide matches
colnames(signatures.cosine)[2] <- "genome.match"
manhattan <- left_join(manhattan, signatures.cosine, by=c("cellid","genome.match"))
#merge cosmic matches
colnames(signatures.cosine) <- c("cellid","cosmic.match","cosmic.cos.sim")
manhattan <- left_join(manhattan, signatures.cosine, by=c("cellid","cosmic.match"))
#merge 18k genes matches
colnames(signatures.cosine) <- c("cellid","gene.match","gene.cos.sim")
manhattan <- left_join(manhattan, signatures.cosine, by=c("cellid","gene.match"))

manhattan$cosmic.lower.cos <- manhattan$cos.sim>manhattan$cosmic.cos.sim
manhattan$gene.lower.cos <- manhattan$cos.sim>manhattan$gene.cos.sim

print("Number of discordant matches between genome-wide and 18k-gene-level resolutions with lower cosine similarity of CNSig activities") 
print(paste0(nrow(manhattan[manhattan$gene.equal==FALSE & manhattan$gene.lower.cos==TRUE,]),"/",322," samples"))

print("Number of discordant matches between genome-wide and cosmic-gene-level resolutions with lower cosine similarity of CNSig activities") 
print(paste0(nrow(manhattan[manhattan$cosmic.equal==FALSE & manhattan$cosmic.lower.cos==TRUE,]),"/",347," samples"))
```

Results limited to ovarian cancer cell lines
```{r}
cells.tissue<-as.character(na.omit(CCLE_metadata$Cell.line.primary.name[CCLE_metadata$Site.Primary=="ovary"]))
ov.manhattan <- manhattan[manhattan$cellid%in%cells.tissue,]

print("Number of discordant HGSOV matches between genome-wide and 18k-gene-level resolutions with lower cosine similarity of CNSig activities") 
print(paste0(nrow(ov.manhattan[ov.manhattan$gene.equal==FALSE & ov.manhattan$gene.lower.cos==TRUE,]),"/",25," samples"))

print("Number of discordant HGSOV matches between genome-wide and cosmic-gene-level resolutions with lower cosine similarity of CNSig activities") 
print(paste0(nrow(ov.manhattan[ov.manhattan$cosmic.equal==FALSE & ov.manhattan$cosmic.lower.cos==TRUE,]),"/",25," samples"))
```


