---
title: "Comparing cell lines using other data"
author:
  - Barbara Hernando, bhernando@cnio.es
  - Blas Chaves, bchavesu@cnio.es
  - Geoff Macintyre, gmacintyre@cnio.es
date: "`r format(Sys.time(), '%d %B, %Y')`"
output:
  html_document:
    number_sections: no
    theme: simplex
    toc: yes
    toc_float: yes
    df_print: paged
---

```{r setup, message=FALSE}
knitr::opts_chunk$set(collapse = TRUE,comment = "#>")
knitr::opts_chunk$set(echo = TRUE)
knitr::opts_chunk$set(message = FALSE)
knitr::opts_chunk$set(warning = FALSE)

## PACKAGES
library(dplyr)
library(CNpare)
library(qusage)

## DATA DIRECTORY
path_to_data<-paste0(dirname(rstudioapi::getSourceEditorContext()$path),"/data/")
```


This Rmarkdown shows the benchmarking of **CNpare tool** against other approaches for choosing matching cell lines.

In order to benchmark CNpare against other approaches for choosing matching cell lines, we computed a number of variables representative of these alternative methods.

### 1. Comparing cell lines using other copy-number data

##### [Input data]{.ul}

```{r cn_input}
cells_segcn <- CNpare::cells_segcn
mapping <- CNpare::cells_mapping
```

```{r data_process}
cells_segcn<-getCINProfiles(segcn=cells_segcn, samples=unique(cells_segcn$sample))

#cell lines with CN profile generated with CCLE and GDSC data
ccle <- mapping[mapping$study=="CCLE",c(1,4)]
ccle <- ccle[ccle$fileid %in% cells_segcn$sample,] #with ASCAT CN profile
gdsc <- mapping[mapping$study=="GDSC",c(1,4)] #in GDSC cellid and fileid is the same
gdsc <- gdsc[gdsc$fileid %in% cells_segcn$sample,] #with ASCAT CN profile
both <- ccle[ccle$cellid %in% gdsc$fileid,] 
```

#### 1.1. Chromosome arm copy number

First, we get the canonical lengths of chromosomal arms
```{r arm_lengths}
## Canonical lengths of chromosomal arms
armlens=data.table::fread(paste0(path_to_data,"/hg19.cytoBand.txt.gz"), 
              col.names = c("chromosome","start","end","name","gieStain"))
armlens=armlens[ , .(start=min(start), end = max(end), length = sum(end - start)), 
                 by = .(chromosome, arm = substring(name, 1, 1)) ]
armlens$chromosome = substr(armlens$chromosome, 4, stop = 6)
armlens$chromosome = factor(armlens$chromosome, levels=c(1:22,"X", "Y"))
armlens=armlens[order(armlens$chromosome),]
armlens$chromosome = as.character(armlens$chromosome)
armlens=armlens[armlens$chromosome!="Y",]
```

We computed the whole arm chromosome copy number by averaging the values of all segments aligned to each chromosome arm. Then, we performed a Pearson correlation to compare CCLE and GDSC cell lines. The propotion of cells matched was then calculated

```{r arm_cn}
## Add chromosome arm in cn segments
chroms <- as.character(unique(cells_segcn$chromosome))
cn <- c()
for (chr in chroms){
    arm <- armlens[armlens$chromosome==chr,]
    c <- cells_segcn[cells_segcn$chromosome==chr,]
    c$arm[c$end<arm$end[1]] <- arm$arm[1]
    c$arm[c$end>=arm$end[1]] <- arm$arm[2]
    
    cn <- rbind(cn, c)
}
cn <- cn[order(cn$sample,cn$chromosome,cn$arm),]

#ccle
samples=unique(both$fileid)
ccle <- matrix(nrow = nrow(armlens), ncol = length(samples)) #Create the matrix
colnames(ccle) <- samples
rownames(ccle) <- paste0(armlens$chromosome, armlens$arm)

for (a in 1:nrow(armlens)){
    chrom <- armlens[a,chromosome]
    arm <- armlens[a,arm]
    
    dat <- cn[(cn$chromosome %in% chrom & cn$arm %in% arm), ]
    for (s in 1:length(samples)){
      ploidy <- mean(cells_segcn[cells_segcn$sample==samples[s],"segVal"])
      abs <- dat[dat$sample==samples[s], "segVal"]
      rel <- log2(abs/ploidy)
      ccle[a,s]<- mean(rel)
    }
}

#gdsc
samples=unique(both$cellid)
gdsc <- matrix(nrow = nrow(armlens), ncol = length(samples)) #Create the matrix
colnames(gdsc) <- samples
rownames(gdsc) <- paste0(armlens$chromosome, armlens$arm)

for (a in 1:nrow(armlens)){
    chrom <- armlens[a,chromosome]
    arm   <- armlens[a,arm]
    
    dat <- cn[(cn$chromosome %in% chrom & cn$arm %in% arm), ]
    for (s in 1:length(samples)){
      ploidy <- mean(cells_segcn[cells_segcn$sample==samples[s],"segVal"])
      abs <- dat[dat$sample==samples[s], "segVal"]
      rel <- log2(abs/ploidy)
      gdsc[a,s]<- mean(rel)
    }
}


#Mask segments that are not gained/lost respect to the overall ploidy (relCN between -0.1 and 0.1)
#Ben-David et al. Nature 560, 325â€“330 (2018). https://doi.org/10.1038/s41586-018-0409-3
#we cannot remove them for being able to create the matrix (equal number of rows/columns needed)
ccle[ccle>-0.1 & ccle<0.1] <- NA
gdsc[gdsc>-0.1 & gdsc<0.1] <- NA

## Pearson correlations
measures<-getSimilarities(dat1=ccle, dat2=gdsc, method="pearson")
colnames(measures)<-c("fileid", "id","r")
measures<-left_join(measures,both,by="fileid")
list.tophits<-getTopHit(samples=both$cellid, measure=measures[,c(4,2:3)], method="pearson")

## Frequency of true matches
list.tophits$equal<-list.tophits[,1]==list.tophits[,2]
top1<-list.tophits[list.tophits$equal,] 
print("Number of samples matching based on arm-level copy number") 
print(paste0(nrow(top1),"/304 samples"))
```

#### 1.2. Ploidy status

For each cell line, we computed the weighted mean copy number values of all segments to infer the overall ploidy. The relative weight of each segment depends on their length. Mean values were then rounded to the closest integer value, and the proportion of cell pairs matched by the ploidy status was then computed
```{r ploidy}
#ccle
samples=unique(both$fileid)
ccle <- c() #Create the matrix

for (s in 1:length(samples)){
    c <- c()
    d <- cells_segcn[cells_segcn$sample==samples[s], ]
    d$length <- d$end-d$start
    ploidy <- stats::weighted.mean(d$segVal, d$length)
    c <- cbind(fileid=samples[s], ploidy=ploidy)
    ccle <- rbind(ccle,c)
}

#gdsc
samples=unique(both$cellid)
gdsc <- c() #Create the matrix

for (s in 1:length(samples)){
    c <- c()
    d <- cells_segcn[cells_segcn$sample==samples[s], ]
    d$length <- d$end-d$start
    ploidy <- stats::weighted.mean(d$segVal, d$length)
    c <- cbind(cellid=samples[s], ploidy=ploidy)
    gdsc <- rbind(gdsc,c)
}

ccle<-as.data.frame(ccle)
ccle$ploidy<-round(as.numeric(ccle$ploidy),1)
gdsc<-as.data.frame(gdsc)
gdsc$ploidy<-round(as.numeric(gdsc$ploidy),1)

ccle<-left_join(ccle,both,by="fileid")
cells<-left_join(ccle,gdsc,by="cellid")
cells$equal<-cells[,2]==cells[,4]

## Frequency of true matches
match<-cells[cells$equal,] 
print("Number of samples matching based on ploidy status") 
print(paste0(nrow(match),"/304 samples"))
```

#### 1.3. Gene-level copy number

Gene positions were downloaded from UCSC Table Browser. We extracted the copy numbers in the transcription start of 18,944 protein-coding genes, and then compared gene-level copy numbers across cell lines using Pearson correlation. We also performed the analysis including only de Cancer Gene Consensus (COSMIC)

First, we get gene positions
```{r}
#get COSMIC genes
cosmic_genes = read.table(paste0(path_to_data,"COSMICgenes.txt"), fill = TRUE, sep="\t")[,1]

#gene positions
genes = read.table(paste0(path_to_data,"knownGenes_positions.txt"),sep="\t")[,c(1:4,7)]
genes = genes[genes$V7 %in% cosmic_genes,] #only oncogenic genes
genes = genes[!duplicated(genes$V3),] #no duplicates in start position
genes = genes[!duplicated(genes$V4),] #only one per protein name
genes = genes[!duplicated(genes$V7),] #only one per gene name --> randomly because positions are similar & we are using segmented tables to get the CN values

pos.genes<-genes[,c(5,2,3,3)] #column 3 txStart
colnames(pos.genes)<- c("name", "chromosome", "start", "end")
pos.genes$chromosome = substr(pos.genes$chromosome, 4, stop = 5)
chr<-c(1:22, "X")
pos.genes<-data.table::as.data.table(pos.genes[pos.genes$chromosome %in% chr,]) #remove genes in Y
```

Using segment tables
```{r}
#ccle
samples=unique(both$fileid)
names=pos.genes$name
ccle <- c()
for (s in 1:length(samples)){
    dat<-data.table::as.data.table(cells_segcn[cells_segcn$sample==samples[s],])
    ploidy=mean(dat$segVal)
    dat<-dat[,c(5,4,1:3)]
    data.table::setkey(pos.genes, chromosome, start, end)
    overlap <- data.table::foverlaps(dat,pos.genes)
    overlap <- overlap[!is.na(overlap$name),]
    
    ccle<-rbind(ccle,cbind(overlap[,c(2,5,6)],ploidy))
}

#gdsc
samples=unique(both$cellid)
names=pos.genes$name
gdsc <- c()
for (s in 1:length(samples)){
    dat<-data.table::as.data.table(cells_segcn[cells_segcn$sample==samples[s],])
    ploidy=mean(dat$segVal)
    dat<-dat[,c(5,4,1:3)]
    data.table::setkey(pos.genes, chromosome, start, end)
    overlap <- data.table::foverlaps(dat,pos.genes)
    overlap <- overlap[!is.na(overlap$name),]
    
    gdsc<-rbind(gdsc,cbind(overlap[,c(2,5,6)],ploidy))
}

#Transform to relative copy-number 
ccle$relCN <- log2(ccle$segVal/ccle$ploidy)
gdsc$relCN <- log2(gdsc$segVal/gdsc$ploidy)

#Mask segments that are not gained/lost respect to the overall ploidy (relCN between -0.3 and 0.3)
#https://depmap.org/portal/cell_strainer/
#we cannot remove them for being able to create the matrix (equal number of rows/columns needed)
ccle$relCN[ccle$relCN>-0.3 & ccle$relCN<0.3] <- NA
gdsc$relCN[gdsc$relCN>-0.3 & gdsc$relCN<0.3] <- NA

mccle<-reshape2::acast(ccle, name~sample, value.var="relCN")
mgdsc<-reshape2::acast(gdsc, name~sample, value.var="relCN")

#pearson correlations
measures<-getSimilarities(dat1=mccle, dat2=mgdsc, method="pearson")
colnames(measures)<-c("fileid", "id","r")
measures<-left_join(measures,both,by="fileid")
list.tophits<-getTopHit(samples=both$cellid, measure=measures[,c(4,2:3)], method="pearson")

#match
list.tophits$equal<-list.tophits[,1]==list.tophits[,2]
top1<-list.tophits[list.tophits$equal,] 
print("Number of samples matching by gene-level copy numbers") 
print(paste0(nrow(top1),"/304 samples"))
```

### 2. Comparing cell lines using gene expression data

##### [Input data]{.ul}

We first download the input data from: https://depmap.org/portal/download/

    - `rnaseq_broad_20210317.csv` -- FPKM and read counts gene expression data from CCLE cell lines
    - `rnaseq_sanger_20210316.csv` -- FPKM and read counts gene expression data from GDSC
    - `model_list_20210719.csv` -- Data with the information of all cell lines
  
```{r gexp_input, include=FALSE}
broad_data<-read.csv(paste0(path_to_data,"rnaseq_broad_20210317.csv")) #CCLE
sanger_data<-read.csv(paste0(path_to_data,"rnaseq_sanger_20210316.csv")) #GDSC
model_id<-read.csv(paste0(path_to_data,"model_list_20210719.csv"))
```

##### [Select cell lines]{.ul}

Then we selected the cell lines that were in both datasets
```{r select_lines}
cells<-model_id$model_id[model_id$model_type=="Cell Line"] #IDs of cell line models
ccle<-unique(broad_data$model_id) #List the cell lines IDs in CCLE
gdsc<-unique(sanger_data$model_id) #List the cell lines IDs in GDSC

models<-c(ccle,gdsc)
both<-duplicated(models)
both<-unique(models[both])
```

##### [Generate matrices]{.ul}

Then we generated the matrices with comparable data, for the CCLE and GDSC data
```{r gexp_matrixes}
ccle_pairs<-broad_data[broad_data$model_id %in% both,]
gdsc_pairs<-sanger_data[sanger_data$model_id %in% both,]

ccle<-c()
gdsc<-c()
for (i in 1:length(both)){
  c<-ccle_pairs[ccle_pairs$model_id==both[i],]
  ccle<-cbind(ccle,c$fpkm)
  
  g<-gdsc_pairs[gdsc_pairs$model_id==both[i],]
  gdsc<-cbind(gdsc,g$fpkm)
}

#Column names 
ccle<-as.data.frame(ccle)
colnames(ccle)<-both
gdsc<-as.data.frame(gdsc)
colnames(gdsc)<-both
```

##### [Pearson correlations & proportion of matched cell lines]{.ul}

```{r matched}
correlations<-getSimilarities(ccle,gdsc,method="pearson")

list.tophits<-getTopHit(samples=both, measure=correlations, method="pearson")
list.tophits$equal<-list.tophits[,1]==list.tophits[,2]
top1<-list.tophits[list.tophits$equal,] 
print("Number of samples matching by gene expression") 
print(paste0(nrow(top1),"/73 samples"))
```

